fairlearn/metrics/_metric_frame.py

```diff
@@ -0,0 +1,515 @@
+
+    def _compute_overall(self, func_dict, y_true, y_pred, cf_list):
+        if cf_list is None:
+            result = pd.Series(index=func_dict.keys(), dtype='object')
+            for func_name in func_dict:
+                metric_value = func_dict[func_name].evaluate_all(y_true, y_pred)
+                result[func_name] = metric_value
+        else:
+            result = self._compute_dataframe_from_rows(func_dict, y_true, y_pred, cf_list)
+        return result
+
+    def _compute_by_group(self, func_dict, y_true, y_pred, sf_list, cf_list):
+        rows = copy.deepcopy(sf_list)
+        if cf_list is not None:
+            # Prepend the conditional features, so they are 'higher'
+            rows = copy.deepcopy(cf_list) + rows
+
+        return self._compute_dataframe_from_rows(func_dict, y_true, y_pred, rows)
+
+    def _compute_dataframe_from_rows(self, func_dict, y_true, y_pred, rows):
+        if len(rows) == 1:
+            row_index = pd.Index(data=rows[0].classes, name=rows[0].name)
+        else:
+            row_index = pd.MultiIndex.from_product([x.classes for x in rows],
+                                                   names=[x.name for x in rows])
+
+        if len(row_index) > _SUBGROUP_COUNT_WARNING_THRESHOLD:
+            msg = _SUBGROUP_COUNT_WARNING.format(len(row_index))
+            logger.warning(msg)
+
+        result = pd.DataFrame(index=row_index, columns=func_dict.keys())
+        for func_name in func_dict:
+            for row_curr in row_index:
+                mask = None
+                if len(rows) > 1:
+                    mask = self._mask_from_tuple(row_curr, rows)
+                else:
+                    # Have to force row_curr to be an unary tuple
+                    mask = self._mask_from_tuple((row_curr,), rows)
+
+                # Only call the metric function if the mask is non-empty
+                if sum(mask) > 0:
+                    curr_metric = func_dict[func_name].evaluate(y_true, y_pred, mask)
+                    result[func_name][row_curr] = curr_metric
+        return result
+
+    @property
+    def overall(self) -> Union[pd.Series, pd.DataFrame]:
+        """Return the underlying metrics evaluated on the whole dataset.
+
+        Returns
+        -------
+        pandas.Series or pandas.DataFrame
+            If no ``control_levels`` were specified, then this is
+            a :class:`pandas.Series` with one row for each underlying metric.
+
+            If ``control_levels`` were specified, then this is a
+            :class:`pandas.DataFrame`, with columns corresponding to the
+            underling metric(s) and the rows indexed by the subgroups of
+            the control features.
+        """
+        return self._overall
+
+    @property
+    def by_group(self) -> pd.DataFrame:
+        """Return the collection of metrics evaluated for each subgroup.
+
+        Returns
+        -------
+        pandas.DataFrame
```

Comment:
I'm now really wondering whether we shouldn't return `pandas.Series` when we have only one metric. This would be consistent with returning just a scalar with `overall`, `difference`, `ratio`, `group_min`, `group_max` in that case as well.

Reply:
That's quite a change you're proposing.

Reply:
i resurfaced this on gitter. i think that there are good usability reasons for the change.

Reply:
Adjusted as agreed at the meeting.